# DATABASE
Here is the processing script of the database I used.

## CIViC
CIViC is an open access, open source, community-driven web resource for Clinical Interpretation of Variants in Cancer.
## DoCM
DoCM, the Database of Curated Mutations, is a highly curated database of known, disease-causing mutations that provides easily explorable variant lists with direct links to source citations for easy verification.

Criteria for inclusion into DoCM:
Clinical evidence
     Drug targets associated with a mutation
     Diagnostic or prognostic markers associated with a mutation
Functional evidence
     Disease function described in cell lines
     Disease function described in animal models
     Extremely recurrent mutation coupled with expert opinion of the significance of the mutation
## ClinVar
ClinVar is a a freely accessible, public archive of reports of the relationships among human variations and phenotypes hosted by the National Center for Biotechnology Information (NCBI) and funded by intramural National Institutes of Health (NIH) funding.
## COSMIC
COSMIC, the Catalogue Of Somatic Mutations In Cancer, is the world's largest and most comprehensive resource for exploring the impact of somatic mutations in human cancer.
## Cancer Biomarker Database
Cancer Biomarkers database, is a resources annotating the biomarkers of tumor response to drugs which currently contains information on 1624 genomic biomarkers of response (sensitivity, resistance, or toxicity) to 310 drugs across 130 types of cancer.
## OncoKB
OncoKB, a Precision Oncology Knowledge Base, includes biological, clinical, and therapeutic information curated from multiple unstructured information resources, including guidelines and recommendations derived from FDA labeling, NCCN guidelines, other disease-specific expert and advocacy group recommendations, and the medical literature.
## CanDL
CanDL, the Cancer Driver Log, is an expert-curated database of potentially actionable driver mutations for molecular pathologists and laboratory directors to facilitate literature-based annotation of genomic testing of tumors. It has mutations with proven functional characterization or that have been targeted clinically or pre-clinically by either existing therapies or investigational agents.
## Kin-Driver
KinDriver,the first database that compiles driver mutations in protein kinases (PKs) with experimental evidence demonstrating their functional role. Kin-driver is a manual expert-curated database that pays special attention to activating mutations (AMs) and can serve as a validation set to develop new generation tools focused on the prediction of gain-of-function driver mutations.

## The IARC TP53 Database
The IARC TP53 Database compiles various types of data and information on human TP53 gene variations related to cancer. Data are compiled from the peer-reviewed literature and from generalist databases.

## ExAC
The Exome Aggregation Consortium (ExAC) is a coalition of investigators seeking to aggregate and harmonize exome sequencing data from a variety of large-scale sequencing projects, and to make summary data available for the wider scientific community. The data set spans 60,706 unrelated individuals sequenced as part of various disease-specific and population genetic studies. The author has removed individuals affected by severe pediatric disease, so this data set should serve as a useful reference set of allele frequencies for severe disease studies. All of the raw data from these projects have been reprocessed through the same pipeline, and jointly variant-called to increase consistency across projects.
## dbCPM
dbCPM is a comprehensive literature-based database of Cancer Passenger Mutations, which contains 941 experimentally supported and 978 putative passenger mutations derived from a manual curation of the literature.

## dbNSFP
dbNSFP is a database developed for functional prediction and annotation of all potential non-synonymous single-nucleotide variants (nsSNVs) in the human genome.
MetaSVM, MetaLR, CADD, VEST3, PROVEAN, 4× fitCons, fathmm-MKL, DANN, SIFT, 2× Polyphen2, LRT, MutationTaster, MutationAssessor, FATHMM, M_CAP, REVEL, MutPred, Eigen and GenoCanyon) and conservation scores(GERP++, SiPhy, 2× phastCons scores and 2× phyloP scores)
